pkg/provision/ec2.go

```diff
@@ -0,0 +1,190 @@
+			{
+				DeviceIndex:              aws.Int64(int64(0)),
+				AssociatePublicIpAddress: aws.Bool(true),
+				DeleteOnTermination:      aws.Bool(true),
+				Groups:                   []*string{group},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.ec2Provisioner.WaitUntilInstanceStatusOk(&ec2.DescribeInstanceStatusInput{
+		InstanceIds: []*string{runResult.Instances[0].InstanceId},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	code := int64(16)
+	status := ""
+	if runResult.Instances[0].State.Code == &code {
+		status = ActiveStatus
+	}
+
+	return &ProvisionedHost{
+		ID:     *runResult.Instances[0].InstanceId,
+		Status: status,
+	}, nil
+}
+
+func (p *EC2Provisioner) Status(id string) (*ProvisionedHost, error) {
+	var status string
+	s, err := p.ec2Provisioner.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(s.InstanceStatuses) == 0 {
+		return nil, fmt.Errorf("cannot retrieve status for instance: %s", id)
+	}
+
+	fmt.Println(*s.InstanceStatuses[0].InstanceStatus.Status)
+
+	if *s.InstanceStatuses[0].InstanceStatus.Status == "ok" {
+		status = ActiveStatus
+	}
+
+	d, err := p.ec2Provisioner.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(d.Reservations) == 0 {
+		return nil, fmt.Errorf("cannot describe instance: %s", id)
+	}
+
+	return &ProvisionedHost{
+		ID:     id,
+		Status: status,
+		IP:     aws.StringValue(d.Reservations[0].Instances[0].PublicIpAddress),
+	}, nil
+}
+
+func (p *EC2Provisioner) Delete(id string) error {
+	i, err := p.ec2Provisioner.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return err
+	}
+	groups := i.Reservations[0].Instances[0].SecurityGroups
+
+	t, err := p.ec2Provisioner.TerminateInstances(&ec2.TerminateInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Deletion status: %s\n", t.TerminatingInstances[0].CurrentState)
+	err = p.ec2Provisioner.WaitUntilInstanceTerminated(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	fmt.Printf("Instance: %s deleted\n", id)
+
+	for _, group := range groups {
+		fmt.Printf("Deleting security group: %s\n", *group.GroupId)
+		_, err := p.ec2Provisioner.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+			GroupId:   group.GroupId,
+		})
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Security group: %s deleted\n", *group.GroupId)
+	}
+
+	return nil
+}
+
+func (p *EC2Provisioner) securityGroup(port string) (*string, error) {
+	targetPort, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, err
+	}
+
+	group, err := p.ec2Provisioner.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
+		Description: aws.String("inlets security group"),
```

Comment:
What happens if we have two hosts and then you delete only one of them?

Do they need a unique group name / ID?

Reply:
The security group will need a unique id, I will modify it.